src/Functions/firstTruthy.cpp

```diff
@@ -0,0 +1,161 @@
+    {
+        ColumnNumbers args;
+        for (size_t i = 0; i + 1 < number_of_arguments; ++i)
+            args.push_back(i);
+        return args;
+    }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+    {
+        if (arguments.empty())
+            throw Exception(ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH,
+                "Function {} requires at least one argument", getName());
+        size_t max_args = 1024;
+        if (arguments.size() > max_args)
+            throw Exception(ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH,
+                "Function {} requires at most {} arguments, got {}", getName(), max_args, arguments.size());
+
+        if (arguments.size() == 1)
+            return arguments[0];
+
+        return getLeastSupertype(arguments);
+    }
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr & result_type, size_t input_rows_count) const override
+    {
+        if (arguments.empty())
+            return result_type->createColumnConstWithDefaultValue(input_rows_count);
+
+        if (arguments.size() == 1)
+            return castColumn(arguments[0], result_type);
+
+        size_t num_columns = arguments.size();
+
+        auto result_col = result_type->createColumn();
+        result_col->reserve(input_rows_count);
+
+        /// Cast all arguments to the result type
+        /// Use this columns to insert values into the result column
+        std::vector<ColumnPtr> casted_columns;
+        casted_columns.reserve(num_columns);
+        for (const auto & arg : arguments)
+        {
+            auto casted_column = castColumn(arg, result_type);
+            casted_column = casted_column->convertToFullColumnIfConst();
+            casted_column = casted_column->convertToFullColumnIfSparse();
+
+            if (casted_column->getDataType() != result_col->getDataType())
+            {
+                throw Exception(ErrorCodes::LOGICAL_ERROR,
+                    "All arguments must cast to the same type, got {} and {} for result type {}",
+                    casted_column->dumpStructure(), result_col->dumpStructure(), result_type->getName());
+            }
+
+            casted_columns.push_back(std::move(casted_column));
+        }
+
+        for (size_t row = 0; row < input_rows_count; ++row)
+        {
+            bool found = false;
+
+            /// Check each argument for truthiness
+            for (size_t arg_idx = 0; !found && arg_idx < num_columns; ++arg_idx)
+            {
+                /// A value is considered "falsey" if it's NULL or the default value for its type
+                /// For example:
+                /// - for numeric types, the default is 0
+                /// - for strings, the default is ''
+                /// - for arrays, the default is []
+                if (!arguments[arg_idx].column->isNullAt(row) && !arguments[arg_idx].column->isDefaultAt(row))
```

Comment:
Null should be the default for a nullable column; probably, the first check is not needed.

Also, 0 is not a default for Nullable column, but it's not "truthy" in my opinion.
Maybe rename this function to `firstNotDefault`?